### PR TITLE
Avoid running Yoga layout on subtrees with RequireNativeLayout

### LINKED ISSUE #7524: ViewManagerBase overwrites layout properties for community modules
1. View manager implements `IViewManagerRequiresNativeLayout` and creates a BitmapImage, e.g. like this:
```cpp
Controls::BitmapIcon b;
b.Width(50);
b.Height(50);
b.UriSource(Windows::Foundation::Uri{ L"https://microsoft.github.io/react-native-windows/img/homepage/cross-platform.png" });
```
At this point the image has the right size.

2. Then ViewManagerBase in RNW goes to do layout and runs through this logic in ViewManagerBase::SetLayoutProps:
https://github.com/microsoft/react-native-windows/blob/ce85295fb12b5762b2eecd79635dd86f39747edd/vnext/Microsoft.ReactNative/Views/ViewManagerBase.cpp#L313-L336

where both `width` and `height` are NaN, so the width/height values are reset away from what the app specified. 
These values are coming from Yoga, see `NativeUIManager::DoLayout`:
https://github.com/microsoft/react-native-windows/blob/0e5c8d98e6ed398b0b6c6c6875bc1ed3b37522fc/vnext/Microsoft.ReactNative/Modules/NativeUIManager.cpp#L906-L921

### PR DESCRIPTION
**Pull request by asklar**
Fixes #7524 

Scenario: View managers that produce panels (e.g. StackPanel) that must lay out to accommodate their children. When the native layout runs, there isn't a way to sync up the corresponding yoga layout.
This manifests itself as follows: Imagine the VM produces a StackPanel with a BitmapImage in it, and the BitmapImage has width/height set. As far as Yoga knows, the stack panel has not set its width/height, so when the yoga layout pass runs, it erases that BitmapImage width/height.
The VM is also unable to participate in yoga layout because shadow nodes are not part of the API.

The fix is to make sure that when we have a view from a VM with RequireNativeLayout, we lay out that node according to Yoga, but all of its descendants should not have their layout trampled over by Yoga and instead the native layout will be the source of truth.

tl;dr: When an [abi] view manager sets a view's Width / Height, they later get overwritten with NaN by ViewManagerBase because Yoga doesn't know about what values are in the native side.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/7525)

### CODE REVIEW

**Comment by acoates-ms** [comment]
Should we be providing a way for custom viewmanagers to override GetYogaCustomMeasureFunc?

**Comment by asklar** [comment]
> Should we be providing a way for custom viewmanagers to override GetYogaCustomMeasureFunc?

The problem is that VMs can only interact with xaml elements so that is their source of truth for layout, and the UI manager thinks the source of truth is the shadow nodes:

https://github.com/microsoft/react-native-windows/blob/2c5ba22a3cfa4447dd227b4f7274c882e1786fed/vnext/Microsoft.ReactNative/Modules/NativeUIManager.cpp#L913-L921

Maybe the solution is for `ABIViewManager` to override `SetLayoutProps` and if the VM has `RequiresNativeLayout==true` then SetLayoutProps should shuttle the width/height properties from XAML to the shadow node instead. Thoughts @acoates-ms ?

**Comment by NickGerleman** [comment]
@asklar for self measure components, I think Yoga calls back into the framework with the shadow node (connected to XAML element) to provide measurements from the native framework instead of Flexbox.

We use that internally for things like TextBox where we need XAML to measure it for us.,

**Comment by acoates-ms** [comment]
For the issue that this is trying to fix, is the VM already setting RequiresNativeLayout==true?  In that case I would expect that the measure func we provide would already return the same value as the XAML H/W props?  At which point the yoga node should get the same value as XAML?

**Comment by asklar** [comment]
Updated fix after discussion with Andrew

**Review by NickGerleman — APPROVED** [review]

**Review by acoates-ms — APPROVED** [review]
